providers/base/bin/pipewire_test.py

```diff
@@ -0,0 +1,608 @@
+    can apply to a particular process without further modification.
+
+    """
+    def __init__(self, kp, ki, kd, setpoint=0):
+        """ Creates a PID controller with given constants and setpoint.
+
+           Arguments:
+           kp, ki, kd: PID constants, see class description.
+           setpoint: desired output value; calls to input_change with
+                     a process output reading will return a desired change
+                     to the input to attempt matching output to this value.
+        """
+        self.setpoint = setpoint
+        self.kp = kp
+        self.ki = ki
+        self.kd = kd
+        self._integral = 0
+        self._previous_error = 0
+        self._change_limit = 0
+
+    def input_change(self, process_feedback, dt):
+        """ Calculates desired input value change.
+
+            Based on process feedback and time interval (dt).
+        """
+        error = self.setpoint - process_feedback
+        self._integral = self._integral + (error * dt)
+        derivative = (error - self._previous_error) / dt
+        self._previous_error = error
+        input_change = (self.kp * error) + \
+                       (self.ki * self._integral) + \
+                       (self.kd * derivative)
+        if self._change_limit and abs(input_change) > abs(self._change_limit):
+            sign = input_change / abs(input_change)
+            input_change = sign * self._change_limit
+        return input_change
+
+    def set_change_limit(self, limit):
+        """Ensures that input value changes are lower than limit.
+
+           Setting limit of zero disables this.
+        """
+        self._change_limit = limit
+
+
+class VolumeController(object):
+    pw_types = {'input': 'SOURCE', 'output': 'SINK'}
+
+    def __init__(self, type, logger=None):
+        """Initializes the volume controller.
+
+           Arguments:
+           type: either input or output
+
+        """
+        self.type = type
+        self._volume = None
+        self.identifier = None
+        self.logger = logger
+
+    def set_volume(self, volume):
+        if not 0 <= volume <= 100:
+            return False
+        command = [
+            "wpctl",
+            "set-volume",
+            "@DEFAULT_AUDIO_%s@" % (self.pw_types[self.type]),
+            str(int(volume)) + "%",
+        ]
+        self._wpctl_output(command)
+        self._volume = volume
+        return True
+
+    def get_volume(self):
+        return self._volume
+
+    def mute(self, mute):
+        mute = str(int(mute))
+        command = [
+            "wpctl",
+            "set-mute",
+            "@DEFAULT_AUDIO_%s@" % (self.pw_types[self.type]),
+            mute,
+        ]
+        self._wpctl_output(command)
+        return True
+
+    def _wpctl_output(self, command):
+        # This method mainly calls wpctl (hence the name). Since wpctl may
+        # return a failure if the audio layer is not yet initialized, we will
```

Comment:
Is this audio layer a package?

Reply:
`wpctl` is part of `wireplumber`, the session manager of `pipewire`.

Reply:
Then how come it has any chance to return failture if job.pxu already requires packages

Reply:
We couldn't expect that image will install everything correctly and that could be a question for all applications: who or where should handle this kind of error ?

Reply:
You can use `fail-on-resource` flag (see [doc](https://checkbox.readthedocs.io/en/stable/reference/units/job.html))
Test will fail immediately if the required package isn't exist

Reply:
If that error is raised here it means that the script being called is wrong, therefore the resource expression that triggered it was wrong (in a way, we were wrong, we must update checkbox!), so lets not handle it. If you want the script to fail when neither are present @diohe0311  is completely correct, use `fail-on-resource`

Reply:
This is a possible route for a follow up!